Enable PKCS 7

daverodgman opened the pull request:
## Description

Enable PKCS 7 by default. Remove the pre-production warnings, and add a Changelog entry.

The Changelog entry is based on the original entry which was removed in bc5f03da .

## Gatekeeper checklist

- [x] **changelog** provided
- [x] **backport** not required
- [x] **tests** not required

tom-cosgrove-arm commented on ChangeLog.d/pkcs7-parser.txt:
```diff
@@ -0,0 +1,15 @@
+     - Only a single digest algorithm per message is supported.
+     - Only 0 or 1 certificates are supported per message, which must be in
+       X.509 format.
+     - There is no support for certificate-revocation lists.
```
```suggestion
     - There is no support for certificate revocation lists.
```

tom-cosgrove-arm commented on ChangeLog.d/pkcs7-parser.txt:
```diff
@@ -0,0 +1,15 @@
+       must be empty.
+     Many thanks to Daniel Axtens, Nayna Jain, and Nick Child from IBM for
+     contributing this feature. Thanks to Demi-Marie Obenour for contributing
+     various improvements, tests and bug-fixes.
```
```suggestion
     various improvements, tests and bug fixes.
```

tom-cosgrove-arm commented on ChangeLog.d/pkcs7-parser.txt:
```diff
@@ -0,0 +1,15 @@
+     - Only the signed data content type, version 1 is supported.
+     - Only DER encoding is supported.
+     - Only a single digest algorithm per message is supported.
+     - Only 0 or 1 certificates are supported per message, which must be in
```
This sounds clunky (plural certificates is correct for 0 but not for 1). How about

```
     - Certificates must be in X.509 format. A message must have either one
       certificate or no certificates.
```

tom-cosgrove-arm commented on include/mbedtls/pkcs7.h:
```diff
@@ -22,12 +22,7 @@
- * almost always the case in practice).
- *
- * Note: For the time being, this implementation of the PKCS7 cryptographic
+ * Note: For the time being, this implementation of the PKCS #7 cryptographic
  * message syntax is a partial implementation of RFC 2315.
```
If we're going to do `s/PKCS7/PKCS #7/` there are quite a few others. And I also think "cryptographic message syntax" should be title cased

```suggestion
 * Note: For the time being, this implementation of the PKCS #7 Cryptographic
 * Message Syntax is a partial implementation of RFC 2315.
```

tom-cosgrove-arm commented on include/mbedtls/mbedtls_config.h:
```diff
@@ -2862,10 +2862,6 @@
- * validation is incomplete, and handling of malformed inputs may not be robust.
- * The API may change.
- *
  * Enable PKCS7 core for using PKCS7 formatted signatures.
```
Do we want to do this (below) here and elsewhere?
```suggestion
 * Enable PKCS #7 core for using PKCS #7-formatted signatures.
```

tom-cosgrove-arm commented on ChangeLog.d/pkcs7-parser.txt:
```diff
@@ -0,0 +1,15 @@
+Features
+   * Added partial support for parsing the PKCS #7 cryptographic message
+     syntax, as defined in RFC 2315. Currently, support is limited to the
```
```suggestion
   * Added partial support for parsing the PKCS #7 Cryptographic Message
     Syntax, as defined in RFC 2315. Currently, support is limited to the
```

tom-cosgrove-arm commented on ChangeLog.d/pkcs7-parser.txt:
```diff
@@ -0,0 +1,15 @@
+   * Added partial support for parsing the PKCS #7 cryptographic message
+     syntax, as defined in RFC 2315. Currently, support is limited to the
+     following:
+     - Only the signed data content type, version 1 is supported.
```
According to the RFC it's
```suggestion
     - Only the signed-data content type, version 1 is supported.
```

tom-cosgrove-arm reviewed (commented):
Some comments and minor changes

daverodgman commented:
I've gone through `pkcs7.h` and tried to consistently use `PKCS #7`, rather than `PKCS7` etc (and used `mbedtls_pkcs7 structure` when referring to the parsed representation). Also fixed a few instances of `ASN1` and replaced those with `ASN.1`.

tom-cosgrove-arm commented:
DCO failure

daverodgman commented:
Last three commits are test improvements:

Added a fix for a broken test dependency.

In fixing this I noticed that we don't have any tests when `MBEDTLS_RSA_C` is unset as it's a generic dependency of all the PKCS7 tests. Removed this generic dependency and added it where specifically needed; added a test that fails when `MBEDTLS_RSA_C` is unset.

tom-cosgrove-arm commented on include/mbedtls/mbedtls_config.h:
```diff
@@ -2877,7 +2873,7 @@
  *
  * This module is required for the PKCS7 parsing modules.
```
```suggestion
 * This module is required for the PKCS #7 parsing modules.
```

tom-cosgrove-arm commented on include/mbedtls/pkcs7.h:
```diff
@@ -160,21 +155,21 @@ typedef struct mbedtls_pkcs7 {
 
 /**
- * \brief          Initialize pkcs7 structure.
+ * \brief          Initialize mbedtls_pkcs7 structure.
```
Note: line 45 of pkcs7.c says "Initializes the pkcs7 structure.`

daverodgman replied:
I'm mostly concerned about the wording in the public-facing headers - inside `pkcs7.c` it's not so much an issue. I'll fix as I'm updating `mbedtls_config.h` though.

tom-cosgrove-arm commented on include/mbedtls/pkcs7.h:
```diff
@@ -22,17 +22,12 @@
+ *  - The only supported PKCS #7 Signed Data syntax version is version 1
  *  - The RFC specifies support for BER. This implementation is limited to
  *    DER only.
  *  - The RFC specifies that multiple digest algorithms can be specified
```
Some minor comments on the lines following (can't make comments on those lines)

`s/PKCS6/PKCS #6/`

`s/crls/CRLs/` (makes it easier to parse mentally, at least for me) We have a field `no_of_crls`, but that's fine

tom-cosgrove-arm reviewed (dismissed):
LGTM

xkqian commented on include/mbedtls/pkcs7.h:
```diff
@@ -22,27 +22,22 @@
  * message syntax is a partial implementation of RFC 2315.
  * Differences include:
  *  - The RFC specifies 6 different content types. The only type currently
  *    supported in Mbed TLS is the signed data content type.
```
```suggestion
 *    supported in Mbed TLS is the signed-data content type.
```

xkqian commented on include/mbedtls/pkcs7.h:
```diff
@@ -22,27 +22,22 @@
  *  - The RFC specifies that multiple digest algorithms can be specified
  *    in the Signed Data type. Only one digest algorithm is supported in Mbed TLS.
- *  - The RFC specifies the Signed Data type can contain multiple X509 or PKCS6
+ *  - The RFC specifies the Signed Data type can contain multiple X.509 or PKCS #6
```
```suggestion
 *  - The RFC specifies the Signed Data type can contain multiple X.509 or PKCS #6 extended
```

xkqian reviewed (commented):
This looks quite good to me except two nitpick improvements.

xkqian reviewed (approved):
LGTM

tom-cosgrove-arm reviewed (approved):
LGTM
